src/examples/algorithms/AnomalyDetection.cpp

```diff
@@ -0,0 +1,219 @@
+          std::cout << std::endl;
+        }
+
+        std::vector<UInt> tpOutput;
+        if (tmImpl == "TP") {
+          std::vector<Real> rTpOutput(tp.nCells());
+          // The temporal pooler uses Real32, so the vector must be converted again
+          auto tpInput = VectorHelpers::castVectorType<UInt, Real>(spOutput);
+          tp.compute(tpInput.data(), rTpOutput.data(), true, true);
+          // And the result is converted ONCE again to UInts
+          tpOutput = VectorHelpers::castVectorType<Real32, UInt>(rTpOutput);
+        } else {
+          auto sparseData = VectorHelpers::binaryToSparse<UInt>(spOutput); // need to convert dense to sparse representation here
+          tm.compute(sparseData.size(), sparseData.data(), true);
+          set<UInt> both;
+          auto active = tm.getActiveCells();
+          auto pred = tm.getPredictiveCells();
+          both.insert(active.begin(), active.end());
+          both.insert(pred.begin(), pred.end());
+          tpOutput.assign(both.begin(), both.end()); //union, like TP.outputType = both
+        }
+
+        if (DEBUG_LEVEL > 3) {
+          std::cout << "Output of temporal pooler: ";
+          VectorHelpers::print_vector(VectorHelpers::binaryToSparse<UInt>(tpOutput), ",");
+          std::cout << std::endl;
+        }
+
+        Real anScore = anomaly.compute(
+          VectorHelpers::binaryToSparse<UInt>(spOutput),     // Spatial pooler (current active columns)
+          VectorHelpers::binaryToSparse<UInt>(lastTPOutput_), // Temporal pooler (previously predicted columns)
+          0,
+          0
+        );
+        // Save the output of the TP for the next iteration...
+        if (tmImpl == "TP") {
+          lastTPOutput_ = VectorHelpers::cellsToColumns(tpOutput, tp.nCellsPerCol());
+        } else {
+          lastTPOutput_ = VectorHelpers::cellsToColumns(tpOutput, tm.getCellsPerColumn());
+        }
+
+        if (DEBUG_LEVEL > 4) {
+          std::cout << "Normalized TP Output: ";
+          VectorHelpers::print_vector(VectorHelpers::binaryToSparse<UInt>(lastTPOutput_), ",");
+        }
+
+        if(DEBUG_LEVEL > 0) {
+          std::cout << "Input:\t" << x << "\tAnomaly: " << anScore << std::endl;
+        }
+
+        return anScore;
+    }
+    // constructor with default parameters
+    AnomalyDetection(Real inputMin = 0.0, Real inputMax = 100.0, Real inputResolution = 0.1,
+      UInt nCols = 2048, UInt nCells = 4, UInt anomalyWindowSize = 2, std::string tmImpl = "TM") : 
+        encoder{25, inputMin, inputMax, 0, 0, inputResolution, false},
+        sp{std::vector<UInt>{static_cast<UInt>(encoder.getOutputWidth())}, std::vector<UInt>{nCols}},
+        tp{sp.getNumColumns(), nCells, 12, 8, 15, 5, .5, .8, 1.0, .1, .1, 0.0, false, 42, true, false}, //FIXME tp & tm should not be initialized both, but I dont know how to conditionally initialize based on impl.
+        tm({sp.getNumColumns()}, nCells), //FIXME ensure same params TP/TM
```

Comment:
I'm looking for help with the upper FIXME: tp & tm should not be initialized both, but I dont know how to conditionally initialize based on impl.